data/transactions/logic/tracer.go

```diff
@@ -93,7 +96,28 @@ import "github.com/algorand/go-algorand/data/transactions"
 //   │                                                      │
 //   │ > AfterTxnGroup                                      │
 //   └──────────────────────────────────────────────────────┘
+//
+//   Block Lifecycle Graph
+//   ┌──────────────────────────────────────────────────────┐
+//   │ Block Evaluation                                     │
+//   │  ┌────────────────────────────────────────────────┐  │
+//   │  │ > BeforeBlock                                  │  │
+//   │  │                                                │  │
+//   │  │  ┌──────────────────────────────────────────┐  │  │
+//   │  │  │ > Transaction/LogicSig Lifecycle         │  │  │
```

Comment:
> This order was surprising to me, but because LogicSig verification happens before the Txns in the block get evaluated the BeforeBlock call will happen after LogicSig verification.

Is LogicSig verification independent of the LogicSig lifecycle?

Reply:
From what I understand it's dependent on where the Tracer is being run.

The simulation code runs this [check](https://github.com/algorand/go-algorand/blob/master/ledger/simulation/simulator.go#L183) before it actually evaluates the block. It's during this that the logic sig is actually evaluated so you would see the logic sig hooks invoked before `BeforeBlock` since we haven't yet started the `BlockEvaluator` at that point.

When the ledger calls `AddBlock`, if you've attached a tracer [`StartEvaluator`](https://github.com/algorand/go-algorand/blob/master/ledger/eval/eval.go#L1551) will be called first (which calls `BeforeBlock`) regardless of whether or not you intend to verify or not.

I'm not sure why simulate changes this from the actual evaluation code, but for the purposes of the diagram I represented the order of the ledger's evaluation logic.

Reply:
thanks for clarifying, resolving